sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -205,10 +205,14 @@ object ConstantPropagation extends Rule[LogicalPlan] {
     def replaceConstants0(expression: Expression) = expression transform {
       case a: AttributeReference => constantsMap.getOrElse(a, a)
     }
-    condition transform {
-      case e @ EqualTo(_, _) if !predicates.contains(e) => replaceConstants0(e)
-      case e @ EqualNullSafe(_, _) if !predicates.contains(e) => replaceConstants0(e)
-    }
+    splitConjunctivePredicates(condition).map {
+      // Don't replace constants in IsNotNull because InferFiltersFromConstraints may infer
+      // another IsNotNull
```

Comment:
sorry I don't get this. Why we can't optimize `IsNotNull(a)` to `IsNotNull(1)`?

Reply:
For example:
`EqualTo(a, 1) && IsNotNull(a)` => `EqualTo(a, 1) && IsNotNull(1)` => `EqualTo(a, 1) && true` => `EqualTo(a, 1)`
and then `InferFiltersFromConstraints` will infer another `IsNotNull(a)` from `EqualTo(a, 1)`.

Reply:
shall we restrict it to `IsNotNull(Attribute)`?

Reply:
OK.

Reply:
@cloud-fan Do you have more comments?